actors/builtin/market/market_actor.go

```diff
@@ -427,78 +411,139 @@ func (a Actor) OnMinerSectorsTerminate(rt Runtime, params *OnMinerSectorsTermina
+						}
+						if deal.VerifiedDeal {
+							timedOutVerifiedDeals = append(timedOutVerifiedDeals, deal)
+						}
+						return nil
 					}
-					newlySlashed := st.processDealInitTimedOut(rt, dealID)
-					big.Add(slashed, newlySlashed)
-				} else {
-					// All deals must have timed out.
-					rt.Abortf(exitcode.ErrIllegalArgument, "not all deals have timed out: %d", dealID)
+
+					// This should not be able to happen
+					rt.Abortf(exitcode.ErrIllegalState, "invalid deal state, unstarted, not timed out")
+				}
+
+				slashAmount, nextEpoch := st.updatePendingDealState(rt, state, deal, dealID, et, lt, rt.CurrEpoch())
+				if !slashAmount.IsZero() {
+					amountSlashed = big.Add(amountSlashed, slashAmount)
 				}
+
+				if nextEpoch != epochUndefined {
+					Assert(nextEpoch > rt.CurrEpoch())
+
+					// TODO: can we avoid having this field?
+					state.LastUpdatedEpoch = rt.CurrEpoch()
+
+					if err := states.Set(dealID, state); err != nil {
+						rt.Abortf(exitcode.ErrPlaceholder, "failed to get deal: %v", err)
+					}
+
+					updatesNeeded[nextEpoch] = append(updatesNeeded[nextEpoch], dealID)
+				}
+
+				return nil
+			}); err != nil {
+				rt.Abortf(exitcode.ErrIllegalState, "failed to iterate deals for epoch: %s", err)
+			}
+			if err := dbe.RemoveAll(i); err != nil {
+				rt.Abortf(exitcode.ErrIllegalState, "failed to delete deals from set: %s", err)
 			}
 		}
 
-		return &slashed
-	}).(abi.TokenAmount)
+		// NB: its okay that we're doing a 'random' golang map iteration here
+		// because HAMTs and AMTs are insertion order independent, the same set of
+		// data inserted will always produce the same structure, no matter the order
+		for epoch, deals := range updatesNeeded {
+			if err := dbe.PutMany(epoch, deals); err != nil {
+				rt.Abortf(exitcode.ErrIllegalState, "failed to reinsert deal IDs into epoch set: %s", err)
+			}
+		}
 
-	// TODO: award some small portion of slashed to caller as incentive
+		ndbec, err := dbe.Root()
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to get root of deals by epoch set: %s", err)
+		}
+
+		ltc, err := lt.Root()
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to flush locked table: %s", err)
+		}
+		etc, err := et.Root()
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to flush escrow table: %s", err)
+		}
+		st.LockedTable = ltc
+		st.EscrowTable = etc
 
-	// Restore verified dataset allowance for verified clients.
-	for _, deal := range verifiedDeals {
+		st.DealOpsByEpoch = ndbec
+
+		return nil
+	})
+
+	for _, d := range timedOutVerifiedDeals {
 		_, code := rt.Send(
 			builtin.VerifiedRegistryActorAddr,
 			builtin.MethodsVerifiedRegistry.RestoreBytes,
 			&verifreg.RestoreBytesParams{
-				Address:  deal.Client,
-				DealSize: big.NewIntUnsigned(uint64(deal.PieceSize)),
+				Address:  d.Client,
+				DealSize: big.NewIntUnsigned(uint64(d.PieceSize)),
 			},
 			abi.NewTokenAmount(0),
 		)
```

Comment:
@zixuanzh  we should change this RestoreBytes to support batching, taking a list of (client, size) pairs as params.

Reply:
filed an issue for this: https://github.com/filecoin-project/specs-actors/issues/341